Panic with datastore.PropertyLoadSaver

I get a panic with the stack trace shown below with "nds". It seems that "x*Record" is nil in "Load"

====================================
// Lädt einen Datensatz aus dem Datastore
func (x *Record) Load(ps []datastore.Property) error {
	x.props = make([]datastore.Property, 0, len(ps)) <============== line storage.go:165
	x.Created = time.Now()
	x.Modified = x.Created
....

=================================
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x47b88a]

goroutine 31 [running]:
panic(0x8ff760, 0xc82000a070)
	/home/tim/Dokumente/go_appengine/goroot/src/runtime/panic.go:481 +0x3e6
main66787.(*Record).Load(0x0, 0xc820285c70, 0x5, 0x5, 0x0, 0x0)
	storage.go:165 +0x75a
github.com/qedus/nds.setValue(0x944b20, 0xc8201e23a0, 0x196, 0xc820285c70, 0x5, 0x5, 0x0, 0x0)
	/home/tim/Dokumente/gopath/src/github.com/qedus/nds/nds.go:155 +0x1e4
github.com/qedus/nds.loadDatastore(0x7f81cd1d36e8, 0xc820270ae0, 0xc8201df720, 0x1, 0x1, 0x7f81cd1cf0a8, 0x800d60, 0x0, 0x0)
	/home/tim/Dokumente/gopath/src/github.com/qedus/nds/get.go:339 +0x5c7
github.com/qedus/nds.getMulti(0x7f81cd1d36e8, 0xc820270ae0, 0xc8202424d0, 0x1, 0x15, 0x800d60, 0xc820273340, 0x97, 0x0, 0x0)
	/home/tim/Dokumente/gopath/src/github.com/qedus/nds/get.go:164 +0x3cf
github.com/qedus/nds.GetMulti.func1(0x7f81cd1d36e8, 0xc820270ae0, 0xc8202768d0, 0x1, 0x1, 0xc8202768e0, 0x0, 0xc8202424d0, 0x1, 0x15, ...)
	/home/tim/Dokumente/gopath/src/github.com/qedus/nds/get.go:78 +0x19c
created by github.com/qedus/nds.GetMulti
	/home/tim/Dokumente/gopath/src/github.com/qedus/nds/get.go:81 +0x32b

--- a/get.go
+++ b/get.go
@@ -58,7 +58,7 @@ func GetMulti(c context.Context,
 	keys []*datastore.Key, vals interface{}) error {
 
 	v := reflect.ValueOf(vals)
-	if err := checkMultiArgs(keys, v); err != nil {
+	if err := checkKeysValues(keys, v); err != nil {
 		return err
 	}
 

--- a/get_test.go
+++ b/get_test.go
@@ -1266,3 +1266,16 @@ func TestPropertyLoadSaver(t *testing.T) {
 		t.Fatal("expected another value")
 	}
 }
+
+func TestUnsupportedValueType(t *testing.T) {
+	ctx, closeFunc := NewContext(t)
+	defer closeFunc()
+
+	keys := []*datastore.Key{
+		datastore.NewIncompleteKey(ctx, "Entity", nil),
+	}
+	entities := make([]int, 1)
+	if err := nds.GetMulti(ctx, keys, entities); err == nil {
+		t.Fatal("expected unsupported value error")
+	}
+}

--- a/get_test.go
+++ b/get_test.go
@@ -1214,3 +1214,55 @@ func TestGetMultiPaths(t *testing.T) {
 		t.Log("End", test.description)
 	}
 }
+
+type loadSaveStruct struct {
+	Value int64
+}
+
+func (lss *loadSaveStruct) Save() ([]datastore.Property, error) {
+	return []datastore.Property{
+		datastore.Property{
+			Name:     "Val",
+			Value:    lss.Value,
+			NoIndex:  true,
+			Multiple: false,
+		},
+	}, nil
+}
+
+func (lss *loadSaveStruct) Load(properties []datastore.Property) error {
+
+	for _, p := range properties {
+		if p.Name == "Val" {
+			lss.Value = p.Value.(int64)
+		}
+	}
+	return nil
+}
+
+func TestPropertyLoadSaver(t *testing.T) {
+	ctx, closeFunc := NewContext(t)
+	defer closeFunc()
+
+	keys := []*datastore.Key{
+		datastore.NewIncompleteKey(ctx, "Entity", nil),
+	}
+	entities := []*loadSaveStruct{
+		&loadSaveStruct{
+			Value: 23,
+		},
+	}
+	keys, err := nds.PutMulti(ctx, keys, entities)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	entities = make([]*loadSaveStruct, 1)
+	if err := nds.GetMulti(ctx, keys, entities); err != nil {
+		t.Fatal(err)
+	}
+
+	if entities[0].Value != 23 {
+		t.Fatal("expected another value")
+	}
+}

--- a/nds.go
+++ b/nds.go
@@ -137,22 +137,53 @@ func unmarshalPropertyList(data []byte, pl *datastore.PropertyList) error {
 	return gob.NewDecoder(bytes.NewBuffer(data)).Decode(pl)
 }
 
+type valueType int
+
+const (
+	valueTypeInvalid valueType = iota
+	valueTypePropertyLoadSaver
+	valueTypeStruct
+	valueTypeStructPtr
+	valueTypeInterface
+)
+
+func checkValueType(val reflect.Value) valueType {
+
+	valType := val.Type()
+	if reflect.PtrTo(valType).Implements(typeOfPropertyLoadSaver) {
+		return valueTypePropertyLoadSaver
+	}
+
+	switch valType.Kind() {
+	case reflect.Struct:
+		return valueTypeStruct
+	case reflect.Interface:
+		return valueTypeInterface
+	case reflect.Ptr:
+		valType = valType.Elem()
+		if valType.Kind() == reflect.Struct {
+			return valueTypeStructPtr
+		}
+	}
+	return valueTypeInvalid
+}
+
 func setValue(val reflect.Value, pl datastore.PropertyList) error {
 
-	if reflect.PtrTo(val.Type()).Implements(typeOfPropertyLoadSaver) {
+	valType := checkValueType(val)
+
+	if valType == valueTypePropertyLoadSaver || valType == valueTypeStruct {
 		val = val.Addr()
+	}
+
+	if valType == valueTypeStructPtr && val.IsNil() {
+		val.Set(reflect.New(val.Type().Elem()))
 	}
 
 	if pls, ok := val.Interface().(datastore.PropertyLoadSaver); ok {
 		return pls.Load(pl)
 	}
 
-	if val.Kind() == reflect.Struct {
-		val = val.Addr()
-	}
-	if val.Kind() == reflect.Ptr && val.Type().Elem().Kind() == reflect.Struct && val.IsNil() {
-		val.Set(reflect.New(val.Type().Elem()))
-	}
 	return datastore.LoadStruct(val.Interface(), pl)
 }
 

--- a/nds.go
+++ b/nds.go
@@ -71,13 +71,43 @@ func init() {
 	gob.Register(appengine.GeoPoint{})
 }
 
-func checkMultiArgs(keys []*datastore.Key, v reflect.Value) error {
-	if v.Kind() != reflect.Slice {
-		return errors.New("nds: vals is not a slice")
+type valueType int
+
+const (
+	valueTypeInvalid valueType = iota
+	valueTypePropertyLoadSaver
+	valueTypeStruct
+	valueTypeStructPtr
+	valueTypeInterface
+)
+
+func checkValueType(valType reflect.Type) valueType {
+
+	if reflect.PtrTo(valType).Implements(typeOfPropertyLoadSaver) {
+		return valueTypePropertyLoadSaver
 	}
 
-	if len(keys) != v.Len() {
-		return errors.New("nds: keys and vals slices have different length")
+	switch valType.Kind() {
+	case reflect.Struct:
+		return valueTypeStruct
+	case reflect.Interface:
+		return valueTypeInterface
+	case reflect.Ptr:
+		valType = valType.Elem()
+		if valType.Kind() == reflect.Struct {
+			return valueTypeStructPtr
+		}
+	}
+	return valueTypeInvalid
+}
+
+func checkKeysValues(keys []*datastore.Key, values reflect.Value) error {
+	if values.Kind() != reflect.Slice {
+		return errors.New("nds: valus is not a slice")
+	}
+
+	if len(keys) != values.Len() {
+		return errors.New("nds: keys and values slices have different length")
 	}
 
 	isNilErr, nilErr := false, make(appengine.MultiError, len(keys))
@@ -91,25 +121,14 @@ func checkMultiArgs(keys []*datastore.Key, v reflect.Value) error {
 		return nilErr
 	}
 
-	if v.Type() == typeOfPropertyList {
+	if values.Type() == typeOfPropertyList {
 		return errors.New("nds: PropertyList not supported")
 	}
 
-	elemType := v.Type().Elem()
-	if reflect.PtrTo(elemType).Implements(typeOfPropertyLoadSaver) {
-		return nil
+	if ty := checkValueType(values.Type().Elem()); ty == valueTypeInvalid {
+		return errors.New("nds: unsupported vals type")
 	}
-
-	switch elemType.Kind() {
-	case reflect.Struct, reflect.Interface:
-		return nil
-	case reflect.Ptr:
-		elemType = elemType.Elem()
-		if elemType.Kind() == reflect.Struct {
-			return nil
-		}
-	}
-	return errors.New("nds: unsupported vals type")
+	return nil
 }
 
 func createMemcacheKey(key *datastore.Key) string {
@@ -137,40 +156,9 @@ func unmarshalPropertyList(data []byte, pl *datastore.PropertyList) error {
 	return gob.NewDecoder(bytes.NewBuffer(data)).Decode(pl)
 }
 
-type valueType int
-
-const (
-	valueTypeInvalid valueType = iota
-	valueTypePropertyLoadSaver
-	valueTypeStruct
-	valueTypeStructPtr
-	valueTypeInterface
-)
-
-func checkValueType(val reflect.Value) valueType {
-
-	valType := val.Type()
-	if reflect.PtrTo(valType).Implements(typeOfPropertyLoadSaver) {
-		return valueTypePropertyLoadSaver
-	}
-
-	switch valType.Kind() {
-	case reflect.Struct:
-		return valueTypeStruct
-	case reflect.Interface:
-		return valueTypeInterface
-	case reflect.Ptr:
-		valType = valType.Elem()
-		if valType.Kind() == reflect.Struct {
-			return valueTypeStructPtr
-		}
-	}
-	return valueTypeInvalid
-}
-
 func setValue(val reflect.Value, pl datastore.PropertyList) error {
 
-	valType := checkValueType(val)
+	valType := checkValueType(val.Type())
 
 	if valType == valueTypePropertyLoadSaver || valType == valueTypeStruct {
 		val = val.Addr()

--- a/put.go
+++ b/put.go
@@ -28,7 +28,7 @@ func PutMulti(c context.Context,
 	}
 
 	v := reflect.ValueOf(vals)
-	if err := checkMultiArgs(keys, v); err != nil {
+	if err := checkKeysValues(keys, v); err != nil {
 		return nil, err
 	}
 
@@ -100,7 +100,7 @@ func Put(c context.Context,
 
 	keys := []*datastore.Key{key}
 	vals := []interface{}{val}
-	if err := checkMultiArgs(keys, reflect.ValueOf(vals)); err != nil {
+	if err := checkKeysValues(keys, reflect.ValueOf(vals)); err != nil {
 		return nil, err
 	}
 

